L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+        new XdsProtocolNegotiatorFactory(
+          InternalNettyChannelCredentials.toNegotiator(fallback)));
+  }
+}
+```
+
+### Netty ChannelCredentials Processing Example
+
+The API provided here is not part of the design. It is meant as a
+demonstration.
+
+```java
+final class ProtocolNegotiators {
+  private static final EnumSet<> understoodTlsFeatures =
+      EnumSet.noneOf(TlsChannelCredentials.Feature.class);
+
+  public static Result from(ChannelCredentials creds) {
+    if (creds instanceof TlsChannelCredentials) {
+      TlsChannelCredentials tlsCreds = (TlsChannelCredentials) creds;
+      Set<TlsChannelCredentials.Feature> incomprehensible =
+          tlsCreds.incomprehensible(understoodTlsFeatures);
+      if (!incomprehensible.isEmpty()) {
+        return Result.error("TLS features not understood: " + incomprehensible);
+      }
+      return Result.negotiator(tls());
+
+    } else if (creds instanceof InsecureChannelCredentials) {
+      return Result.negotiator(plaintext());
+
+    } else if (creds instanceof CompositeChannelCredentials) {
+      CompositeChannelCredentials compCreds = (CompositeChannelCredentials) creds;
+      return from(compCreds.getChannelCredentials())
+          .withCallCredentials(compCreds.getCallCredentials());
+
+    } else if (creds instanceof NettyChannelCredentials) {
+      NettyChannelCredentials nettyCreds = (NettyChannelCredentials) creds;
+      return Result.negotiator(nettyCreds.getNegotiator());
+
+    } else if (creds instanceof ChoiceChannelCredentials) {
+      ChoiceChannelCredentials choiceCreds = (ChoiceChannelCredentials) creds;
+      StringBuilder error = new StringBuilder();
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        Result result = from(innerCreds);
+        if (!result.isError()) {
+          return result;
+        }
+        error.append("; ");
+        error.append(result.getError());
+      }
+      return Result.error(error.substring(2));
+
+    } else {
+      return Result.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  public final class Result {
+    public static Result error(String error) {...}
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted. There is also currently no known
+use-case for `ChoiceServerCredentials`, so it will be omitted.
+
+```java
+// instead of:
+server = ServerBuilder.forPort(443)
+    .useTransportSecurity(new File("cert.pem"), new File("cert.key"))
+    .build().start();
+server = ServerBuilder.forPort(8080)
+    .build().start();
+// the user would now:
+ChannelCredentials tlsCreds = new TlsServerCredentials(
```

Comment:
FWIW you would need to provide at least as many overloads as SslContextBuilder does.

Reply:
I hope we can avoid that. I think this API will cater to the 80-90% case, so there is a stable API available to them. We would have some way to let people convert (wrap) a Netty SslContext to a ChannelCredential so that there would still be the same advanced-but-unstable API we have today when people call `nettyChannelBuilder.sslContext()` (ditto for server-side). We can have something equivalent for OkHttp to specify the SSLSocketFactory.

So, for example, I totally expect some stuff to support client certificates in the future, which would imply a setting for the server to {NONE, ASK, REQUIRE} client certificates and a setting to specify non-default root certificates for trust (that could be two Features, or one).

But I don't think we'll ever want to allow people to change the ciphers from this API, and we may be able to avoid letting people specify TrustManagers.

Reply:
I am wondering about the statement "...we may be able to avoid letting people specify TrustManagers." Currently `SslContextBuilder` allows specifying a `TrustManager` on the server side so those use-cases (not sure what those are though) could benefit from the `TrustManager`, right?

Reply:
I was saying that we may have use-cases that require a `TrustManager` continue using `SslContextBuilder`. It would be like today; they would use an experimental API. If we upgraded to Netty 5 at some point we might break the API.